components/brave_ads/core/internal/common/time/time_constraint_util.cc

```diff
@@ -5,21 +5,29 @@
 
 #include "brave/components/brave_ads/core/internal/common/time/time_constraint_util.h"
 
-#include "base/ranges/algorithm.h"
 #include "base/time/time.h"
 
 namespace brave_ads {
 
 bool DoesHistoryRespectRollingTimeConstraint(
     const std::vector<base::Time>& history,
     const base::TimeDelta time_constraint,
-    const size_t cap) {
-  const size_t count = base::ranges::count_if(
-      history, [time_constraint](const base::Time created_at) {
-        return base::Time::Now() - created_at < time_constraint;
-      });
+    size_t cap) {
+  const base::Time threshold = base::Time::Now() - time_constraint;
 
-  return count < cap;
+  for (auto iter = history.crbegin(); iter != history.crend(); ++iter) {
+    if (*iter <= threshold) {
+      // If the time point is older than the threshold, break the loop.
+      break;
+    }
+
+    --cap;
```

Comment:
this can lead to insigned underflow if cap value is already 0

Reply:
Great catch, thank you